src/app/config/resolve_addr.c

```diff
@@ -36,6 +61,469 @@ reset_last_resolved_addr_v4(void)
+                    int warn_severity, const bool explicit_ip)
+{
+  tor_assert(addr);
+
+  /* Public address, this is fine. */
+  if (!tor_addr_is_internal(addr, 0)) {
+    goto allow;
+  }
+
+  /* We have a private IP address. It is allowed only if we set custom
+   * directory authorities. */
+  if (using_default_dir_authorities(options)) {
+    log_fn(warn_severity, LD_CONFIG,
+           "Address '%s' is a private IP address. Tor relays that use "
+           "the default DirAuthorities must have public IP addresses.",
+           fmt_addr(addr));
+    return ERR_DEFAULT_DIRAUTH;
+  }
+
+  if (!explicit_ip) {
+    /* Even with custom directory authorities, only an explicit internal
+     * address is accepted. */
+    log_fn(warn_severity, LD_CONFIG,
+           "Address %s was resolved and thus not explicitly "
+           "set. Even if DirAuthorities are custom, this is "
+           "not allowed.", fmt_addr(addr));
+    return ERR_ADDRESS_IS_INTERNAL;
+  }
+
+ allow:
+  return 0;
+}
+
+/** @brief Get IP address from the given config line and for a specific address
+ *         family.
+ *
+ * This can fail is more than two Address statement are found for the same
+ * address family. It also fails if no statement is found.
+ *
+ * On failure, no out parameters should be used or considered valid.
+ *
+ * @param options Global configuration options.
+ * @param warn_severity Log level that should be used on error.
+ * @param family IP address family. Only AF_INET and AF_INET6 are supported.
+ * @param method_out OUT: String denoting by which method the address was
+ *                   found. This is described in the control-spec.txt as
+ *                   actions for "STATUS_SERVER".
+ * @param hostname_out OUT: String containing the hostname gotten from the
+ *                     Address value if any.
+ * @param addr_out OUT: Tor address of the address found in the cline or
+ *                 resolved from the cline.
+ *
+ * @return Return 0 on success that is an address has been found or resolved
+ *         successfully. Return error code ERR_* found at the top of the file.
+ */
+static int
+get_address_from_config(const or_options_t *options, int warn_severity,
+                        int family, const char **method_out,
+                        char **hostname_out, tor_addr_t *addr_out)
+{
+  bool explicit_ip = false;
+  int num_valid_addr = 0;
+
+  tor_assert(options);
+  tor_assert(addr_out);
+  tor_assert(method_out);
+  tor_assert(hostname_out);
+
+  log_debug(LD_CONFIG, "Attempting to get address from configuration");
+
+  if (!options->Address) {
+    log_info(LD_CONFIG, "No Address option found in configuration.");
+    return ERR_NO_OPT_ADDRESS;
+  }
+
+  for (const config_line_t *cfg = options->Address; cfg != NULL;
+       cfg = cfg->next) {
+    int af;
+    tor_addr_t addr;
+
+    af = tor_addr_parse(&addr, cfg->value);
+    if (af == family) {
+      tor_addr_copy(addr_out, &addr);
```

Comment:
Should we set `hostname_out` in this case too?

Reply:
I did set it to NULL per previous review. But in the case it is an address, no `hostname_out` should remain NULL if I correctly read the previous function `resolve_my_address()`.